onnx/onnx.in.proto

```diff
@@ -250,6 +250,11 @@ message GraphProto {
   // Information for the values in the graph. The ValueInfoProto.name's
   // must be distinct. It is optional for a value to appear in value_info list.
   repeated ValueInfoProto value_info = 13;
+  
+  // Functions referred by this graph. A function is a subgraph composed of
+  // primitive ONNX ops. A node may use function graph's name as its op_type
+  // when referring to a function instead of a primitive ONNX op.
+  repeated GraphProto function = 14;
```

Comment:
One thing I'd like to point out is that, we might want to put a reserved word prefix to denote functions, if we want to refer to functions by name. This is for extensibility purposes:

(1) say we do not have an operator "Foo" defined now.
(2) one user defines a model that has a function called Foo.
(3) some time later, we add an operator "Foo" to the onnx operator set.

This would cause the model definition somewhat ambiguous. We should probably explicitly say

(1) functions should have an explicit prefix that shall not be used by operator names, or
(2) functions take precedence when resolving operator types.

Reply:
Thank you very much @Yangqing  for the great comments!

In my personal thinking, there're two kinds of functions, one is common function in ONNX domain, which will be part of ONNX spec, same as all current primitive ops, another is customized function which will have its own domain and also its body defined in a model, like an "operator" declaration defined in a model with body (subgraph). In this way, there should be no name conflict based on different namespace (domain). Conflicts should be detected when registering an common function or adding a customized function into a model. Make sense please?

btw, @ezyang helped me a lot here and kindly suggested me having motivation 2 resolved firstly (https://github.com/onnx/onnx/wiki/Proposal----Adding-Function-into-ONNX). That's why in this PR, only part of changes https://github.com/linkerzhang/onnx/blob/kezhan/add_function_private/onnx/onnx.in.proto is contained. However, after more thinking this thru, I realize that we might still take motivation 1 together into consideration since same mechanism should be designed for the 2 motivation in my opinion, instead of having different ones for each.

Let's sync and understand more on it please.